### PR TITLE
Improve LDAP heartbeat search query and introduced LDAP_HEARTBEAT_SEARCH_QUERY configuration 

### PR DESCRIPTION
**Pull request by siacomuzzi**
Current heartbeat query is too general and might cause timeout issues.
This PR replaces it with a valid default query ( which can be overridden) that returns no results (or in the worst case, it will return a small number of entries).

Changes:
- Configurable heartbeat search query
- Constrained heartbeat search to base dn 

TODO(will be done after this PR merge): Add documentation about new config - https://auth0.com/docs/extensions/ad-ldap-connector/ad-ldap-connector-config-file-schema

### CODE REVIEW

**Comment by rymir** [comment]
LGTM

**Comment by rymir** [comment]
@siacomuzzi when are we planning a release?

File: lib/ldap.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ function initializeConnection () {
   });
 
   connection.heartbeat = function (callback) {
-    connection.search('', '(objectclass=*)', function (err, res) {
+    connection.search('', '(&(objectclass=user)(|(sAMAccountName=foo)(UserPrincipalName=foo)))', function (err, res) {
```
Comment (madsharm) [inline lib/ldap.js:31]: Two suggestions:
- Constraint the search under configured base DN, will make the search faster
- Make heartbeat search query configurable. We can have the above query in PR as default

Comment (siacomuzzi) [reply lib/ldap.js:31]: please go ahead with your suggestions, we have a customer waiting for this change.

File: lib/initConf.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ var defaults = {
   LDAP_SEARCH_GROUPS:                   '(member:1.2.840.113556.1.4.1941:={0})',
   LDAP_USER_BY_NAME:                    '(sAMAccountName={0})',
   LDAP_NUMBER_OF_PARALLEL_BINDS:        1,
+  LDAP_HEARTBEAT_SEARCH_QUERY:          '(&(objectclass=user)(|(sAMAccountName=foo)(UserPrincipalName=foo)))',
```
Comment (madsharm) [inline lib/initConf.js:13]: We just need a valid query which may or may not return result. This query is also used when setting up connector - https://github.com/auth0/ad-ldap-connector/blob/cbff24a2090901d7d5aeccb5406a32ac13aad8dc/admin/test_config.js#L34

Comment (madsharm) [reply lib/initConf.js:13]: `LDAP_HEARTBEAT_SEARCH_QUERY ` config can be overriden in config.json

File: lib/ldap.js [hunk 3]
```diff
@@ -28,7 +28,12 @@ function initializeConnection () {
   });
 
   connection.heartbeat = function (callback) {
-    connection.search('', '(objectclass=*)', function (err, res) {
+    var opts = {
+      scope: 'sub',
+      filter: nconf.get('LDAP_HEARTBEAT_SEARCH_QUERY'),
+      sizeLimit: 1
```
Comment (siacomuzzi) [inline lib/ldap.js:34]: please remove this, remember that it will return an error when query returns more than 1 entries

Comment (madsharm) [reply lib/ldap.js:34]: Right, Thanks for pointing out. I will remove sizeLimit.

**Review by andre-auth0 — APPROVED** [review]
query will be way faster on large ldap directories

**Comment by madsharm** [comment]
The fix is included in new connector release 6.1.1 - https://cdn.auth0.com/connector/windows/latest.json

**Pull request merged**